src/ir/stack-utils.h

```diff
@@ -158,6 +110,9 @@ struct StackSignature {
     return params == other.params && results == other.results &&
            unreachable == other.unreachable;
   }
+
+  // Returns the LUB of `a` and `b`. Assumes that a LUB exists.
+  static StackSignature getLeastUpperBound(StackSignature a, StackSignature b);
```

Comment:
Should it be obvious what an LUB means here? The effects of unreachability in particular are not obvious enough to me without reading the implementation's source code. Maybe I am just missing an intuition for how/where an LUB would be used.

Reply:
I'll add a comment clarifying the definition. This will be used in finalizing a block like this:

```
(block
  (i32.const 0
  (ref.null extern)
  (br 0
    (pop i32 extern)
  )
  (ref.null func)
)
```

In this case the signatures are `[]->[i32, externref]` and `[]->[funcref]{unreachable}` and the LUB would be `[]->[i32, anyref]`.

Reply:
In what situation would this arise? That is, when will we have such a block but not already know its type?

Reply:
Part of validation involves refinalizing a node then checking that it's actual type matches its refinalized type. So even though we already know the type of the node, we have to recalculate it to make sure we are correct about what the type is.

Reply:
Oh, right, we don't use the existing type there.

I think we might be able to, though, if it would make all this simpler?

Reply:
Some basic question about the Binaryen block validation rules that are not necessarily related to the Poppy IR or this PR:
In your example, even if the `br` is an unconditional branch, can you remind me why we still consider the type of the last element in the block?

Reply:
> Oh, right, we don't use the existing type there.
> 
> I think we might be able to, though, if it would make all this simpler?

AFAICT, The validator has no choice to recalculate the type of each expression because it needs to be able to detect expressions with incorrect types. I'm not sure how that could be changed without making the validator insensitive to that kind of typing problem.

> Some basic question about the Binaryen block validation rules that are not necessarily related to the Poppy IR or this PR:
> In your example, even if the `br` is an unconditional branch, can you remind me why we still consider the type of the last element in the block?

Good question! When Binaryen IR is emitted, `unreachable` blocks are unconditionally given block signature `[]->[]`.

https://github.com/WebAssembly/binaryen/blob/cf51ca4f3b727e7efcfc043bb8120ed2c766285d/src/wasm/wasm-stack.cpp#L25-L27

That means that only blocks that could have type `none` can be candidates to be `unreachable`. In contrast, my old local branch has Poppy IR go through a type inference pass that assigns concrete types to all `unreachable` blocks before it is emitted, so before that pass it is ok for blocks of any type to be marked `unreachable`. We could loosen the restrictions on Binaryen IR by running it through a similar type inference pass before emitting it.

Reply:
Sorry I'm not sure if I understand. I was mostly asking about the current AST's validation rule (not Poppy IR), in `Block::finalize`. If a `br` is in the middle of a block, there shouldn't be a chance that we fall through at the end of the block, but should we still consider the type of the last element in computing LUB?

> In contrast, my old local branch has Poppy IR go through a type inference pass that assigns concrete types to all `unreachable` blocks before it is emitted, so before that pass it is ok for blocks of any type to be marked `unreachable`. We could loosen the restrictions on Binaryen IR by running it through a similar type inference pass before emitting it.

Where is your old local branch? Have we seen that type inference pass or have you uploaded it as a PR?

Reply:
Consider what would happen right now if this Binaryen IR block were given type `unreachable`:

```
(block
  (unreachable)
  (i32.const 0)
)
```
If this block were unreachable, the current code in `BinaryInstWriter::emitResultType` would emit it into a binary with signature `[]->[]`. But that would be an invalid Wasm binary because the block returns an `i32`. Binaryen avoids emitting invalid blocks like this by forcing the blocks to have concrete types in Binaryen IR so that they will be emitted with the correct type rather than with `Type::none`.

Reply:
Here is the type inference pass: https://github.com/WebAssembly/binaryen/pull/2784/files#diff-47b86a03ca3d74296cfc57759de43be1cee592a25325824de449a0bf8f026a33R1-R180.